### DISCUSSION TITLE
thread sanitizer (tsan) data races with `.send()`

### DISCUSSION THREAD

**Discussion by jaredh159** [pointfreeco/swift-composable-architecture#2072] [Q&A]
i have some code in a reducer similar to this example from a recent episode:

```swift
case .loadResponse:
  state.isLoading = false
  return .send(.delegate(.goToCounter(state.count)))
```

where a reducer uses `.send()` to immediately feed an effect back into the system. i always develop with thread sanitizer enabled, and i started getting fairly consistent (maybe 80% of the time) access race warnings about multiple threads accessing the view stores send functionality.

when i changed the code to this:

```swift
case .loadResponse:
  state.isLoading = false
  return .run { send in
    await send(.delegate(.goToCounter(state.count)))
  }
```

then the tsan warnings went away.

now... my app is really complex with a bunch of hairy entitlements, and it's not open source, so it's difficult to show you guys a minimal repro. i'm willing to try to isolate it and reproduce it, but i thought before i went through the effort to do that i would ask a couple things:

1. is this a known issue at all?
2. do you guys test in CI or develop with TSAN enabled?
3. could it be a false positive?
4. is it more likely to be something about the surrounding code that i'm doing wrong?
5. would a stack trace or more info about the TSAN warning be helpful on it's own, independent of a full repro?

thanks so much!

EDIT: oh, forgot to mention, i'm running the very latest commit of the `prerelease/1.0` branch, fwiw

**Comment by mbrandonw** [comment]
Hi @jaredh159, some answers to your questions:

> is this a known issue at all?

It is not, as far as we know.

> do you guys test in CI or develop with TSAN enabled?

We do sometimes, and haven't come across this issue. I also just turned it on for the demo we have been building in episodes and wasn't able to reproduce.

> could it be a false positive?

I personally doubt that.

> is it more likely to be something about the surrounding code that i'm doing wrong?

That is a possibility. The only thing that comes to mind is if you somehow have an effect that emits or completes on the non-main thread. This is really only possible when using Combine-based effects rather than async effects, so if you have any of those that would be where to look.

> would a stack trace or more info about the TSAN warning be helpful on it's own, independent of a full repro?

That could help, or if you are able to reproduce the problem in any of our demo apps.

**Reply by jaredh159** [reply]
thanks for the reply @mbrandonw that's helpful. i don't have any combine-based effects in that area of the app, but i dug around a bit more, and will continue to investigate and possibly work on a full repro, or a repro with a demo app, if i can pull it off.
